lib/portus/ldap.rb

```diff
@@ -0,0 +1,141 @@
+require "net/ldap"
+require "devise/strategies/authenticatable"
+
+module Portus
+  # Portus::LDAP implements Devise's authenticatable for LDAP servers. This
+  # class will fallback to other strategies if the given user does not exist
+  # in the LDAP server or LDAP is disabled.
+  #
+  # If the given user could be found, it will try to authenticate it through
+  # the LDAP server. If we can bind to the server with the given credentials,
+  # we assume that the authentication was successful. In this case, if this
+  # is the first time that this user enters Portus, it will be saved inside of
+  # Portus' DB. There are some issues while doing this:
+  #
+  #   1. The 'email' is not provided in a standard way: some LDAP servers may
+  #      provide it, some others won't. Therefore, once an LDAP user logs in
+  #      Portus for the first time, the email will be left blank. This should
+  #      be handled by the controller layer.
+  #   2. The 'password' is stored in the DB but it's not really used. This is
+  #      because the DB requires the password to not be blank, but in order to
+  #      authenticate we always want to check with the LDAP server.
+  #
+  # This class is only useful if LDAP is enabled in the `config/config.yml`
+  # file. Take a look at this file in order to read more on the different
+  # configurable values.
+  class LDAP < Devise::Strategies::Authenticatable
+    # Re-implemented from Devise::Strategies::Authenticatable to authenticate
+    # the user.
+    def authenticate!
+      ldap = load_configuration
+
+      # If LDAP is enabled and the user actually exists, try to authenticate
+      # through the LDAP server. Otherwise we fall back to the next strategy.
+      if ldap && ldap.search(search_options)
+        # The user can be located in the LDAP server, try to authenticate
+        # through it. If there's any failure, the authentication process will
+        # fail without going to the any other strategy.
+        if ldap.bind_as(search_options.merge(password: password))
+          user = find_or_create_user!
+          user.valid? ? success!(user) : fail!(:invalid_login)
+        else
+          fail!(:invalid_login)
+        end
+      else
+        # rubocop:disable Style/SignalException
```

Comment:
Being there means:
  1. ldap is disabled
  2. ldap is enabled, but the user cannot be found on the remote server

The next action is `fail` which moves to the next authentication strategy (portus' db). I think it would be better to distinguish the second case and fail immediately. I think that once LDAP is enable we should not allow a mixed authentication strategy.

Did I get something wrong?